Tests/NIOPosixTests/AsyncChannelBootstrapTests.swift

```diff
@@ -278,9 +280,9 @@ final class AsyncChannelBootstrapTests: XCTestCase {
 
             group.addTask {
                 try await withThrowingTaskGroup(of: Void.self) { group in
-                    for try await childChannel in channel.inboundStream {
+                    for try await negotiationResult in channel.inboundStream {
                         group.addTask {
-                            switch childChannel {
+                            switch try await negotiationResult.resolve() {
```

Comment:
This part here is the important change. Instead of having a `NIOAsyncChannel` of negotiation results we now have a channel of futures of negotiation results. This means that users have to resolve the result and can then decide what to do with it. It doesn't look too bad from a users perspective and allows us to significantly reduce the amount of code in NIO. cc @Lukasa @glbrntt @rnro